// ngx-gtag/esm2015/lib/google-anaytics/google-anaytics.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser, DOCUMENT } from '@angular/common';
import { Router, NavigationEnd } from '@angular/router';
import { filter } from 'rxjs/operators';
import { NGX_GTAG_CONFIG } from '../google-analytics.token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../google-analytics.token";
import * as i3 from "@angular/common";
export class GoogleAnayticsService {
    /**
     * @param {?} router
     * @param {?} config
     * @param {?} document
     * @param {?} platform
     */
    constructor(router, config, document, platform) {
        this.router = router;
        this.config = config;
        this.document = document;
        this.platform = platform;
    }
    /**
     * @return {?}
     */
    startTracking() {
        if (isPlatformBrowser(this.platform)) {
            if (Array.isArray(this.config)) {
                this.config.forEach((conf) => {
                    conf.options = {};
                    /** @type {?} */
                    let s1 = this.renderer.createElement('script');
                    s1.src = `https://www.googletagmanager.com/gtag/js?id='${conf.trackingId}'`;
                    this.renderer.appendChild(this.document.head, s1);
                });
            }
            else if (this.config.options === undefined) {
                this.config.options = {};
                /** @type {?} */
                let s1 = this.renderer.createElement('script');
                s1.src = `https://www.googletagmanager.com/gtag/js?id='${this.config.trackingId}'`;
                this.renderer.appendChild(this.document.head, s1);
            }
            /** @type {?} */
            let s2 = this.renderer.createElement('script');
            /** @type {?} */
            const text = this.renderer.createText(`window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments);}gtag('js',new Date());`);
            this.renderer.appendChild(s2, text);
            this.renderer.appendChild(this.document.head, s2);
            this.router.events
                .pipe(filter(event => event instanceof NavigationEnd))
                .subscribe((navEnd) => {
                if (Array.isArray(this.config)) {
                    this.config.forEach((conf) => {
                        conf.options.page_path = navEnd.urlAfterRedirects;
                        gtag('config', conf.trackingId, conf.options);
                    });
                }
                else {
                    this.config.options.page_path = navEnd.urlAfterRedirects;
                    gtag('config', this.config.trackingId, this.config.options);
                }
            });
        }
    }
}
GoogleAnayticsService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
GoogleAnayticsService.ctorParameters = () => [
    { type: Router },
    { type: undefined, decorators: [{ type: Inject, args: [NGX_GTAG_CONFIG,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
/** @nocollapse */ GoogleAnayticsService.ngInjectableDef = i0.defineInjectable({ factory: function GoogleAnayticsService_Factory() { return new GoogleAnayticsService(i0.inject(i1.Router), i0.inject(i2.NGX_GTAG_CONFIG), i0.inject(i3.DOCUMENT), i0.inject(i0.PLATFORM_ID)); }, token: GoogleAnayticsService, providedIn: "root" });
if (false) {
    /** @type {?} */
    GoogleAnayticsService.prototype.trackingId;
    /** @type {?} */
    GoogleAnayticsService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.document;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.platform;
}
