// ngx-gtag/esm5/lib/google-analytics-events/google-analytics-events.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Directive, Renderer2, Input, ElementRef } from '@angular/core';
import { NgxGtagEventService } from './google-analytics-events.service';
var GoogleAnalyticsEventsDirective = /** @class */ (function () {
    function GoogleAnalyticsEventsDirective(ngxGtagEventService, renderer, el) {
        this.ngxGtagEventService = ngxGtagEventService;
        this.renderer = renderer;
        this.el = el;
    }
    /**
     * @return {?}
     */
    GoogleAnalyticsEventsDirective.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        try {
            this.renderer.listen(this.el.nativeElement, this.gtagOn, function () {
                _this.ngxGtagEventService.event({ action: _this.gtagAction || _this.gtagOn, options: tslib_1.__assign({}, _this.gtagOptions) });
            });
        }
        catch (err) {
            console.error(err);
        }
    };
    GoogleAnalyticsEventsDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ngxGtagEvent]'
                },] }
    ];
    /** @nocollapse */
    GoogleAnalyticsEventsDirective.ctorParameters = function () { return [
        { type: NgxGtagEventService },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    GoogleAnalyticsEventsDirective.propDecorators = {
        gtagOn: [{ type: Input }],
        gtagTrackingId: [{ type: Input }],
        gtagAction: [{ type: Input }],
        gtagOptions: [{ type: Input }]
    };
    return GoogleAnalyticsEventsDirective;
}());
export { GoogleAnalyticsEventsDirective };
if (false) {
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagOn;
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagTrackingId;
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagAction;
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagOptions;
    /**
     * @type {?}
     * @private
     */
    GoogleAnalyticsEventsDirective.prototype.ngxGtagEventService;
    /**
     * @type {?}
     * @private
     */
    GoogleAnalyticsEventsDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    GoogleAnalyticsEventsDirective.prototype.el;
}
