// ngx-gtag/esm2015/lib/google-analytics-events/google-analytics-events.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Renderer2, Input, ElementRef } from '@angular/core';
import { NgxGtagEventService } from './google-analytics-events.service';
export class GoogleAnalyticsEventsDirective {
    /**
     * @param {?} ngxGtagEventService
     * @param {?} renderer
     * @param {?} el
     */
    constructor(ngxGtagEventService, renderer, el) {
        this.ngxGtagEventService = ngxGtagEventService;
        this.renderer = renderer;
        this.el = el;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        try {
            this.renderer.listen(this.el.nativeElement, this.gtagOn, () => {
                this.ngxGtagEventService.event({ action: this.gtagAction || this.gtagOn, options: Object.assign({}, this.gtagOptions) });
            });
        }
        catch (err) {
            console.error(err);
        }
    }
}
GoogleAnalyticsEventsDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxGtagEvent]'
            },] }
];
/** @nocollapse */
GoogleAnalyticsEventsDirective.ctorParameters = () => [
    { type: NgxGtagEventService },
    { type: Renderer2 },
    { type: ElementRef }
];
GoogleAnalyticsEventsDirective.propDecorators = {
    gtagOn: [{ type: Input }],
    gtagTrackingId: [{ type: Input }],
    gtagAction: [{ type: Input }],
    gtagOptions: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagOn;
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagTrackingId;
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagAction;
    /** @type {?} */
    GoogleAnalyticsEventsDirective.prototype.gtagOptions;
    /**
     * @type {?}
     * @private
     */
    GoogleAnalyticsEventsDirective.prototype.ngxGtagEventService;
    /**
     * @type {?}
     * @private
     */
    GoogleAnalyticsEventsDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    GoogleAnalyticsEventsDirective.prototype.el;
}
