// ngx-gtag/esm5/lib/google-analytics.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { NGX_GTAG_CONFIG } from './google-analytics.token';
import { GoogleAnayticsComponent } from './google-anaytics/google-anaytics.component';
import { GoogleAnalyticsEventsDirective } from './google-analytics-events/google-analytics-events.directive';
var NgxGtagModule = /** @class */ (function () {
    function NgxGtagModule(parentModule) {
        if (parentModule) {
            throw new Error('ngx-gtag should only be loaded in one module, preferably your root or core module.');
        }
    }
    /**
     * @param {?} options
     * @return {?}
     */
    NgxGtagModule.forRoot = /**
     * @param {?} options
     * @return {?}
     */
    function (options) {
        return {
            ngModule: NgxGtagModule,
            providers: [
                {
                    provide: NGX_GTAG_CONFIG,
                    useValue: options
                }
            ]
        };
    };
    NgxGtagModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        GoogleAnayticsComponent,
                        GoogleAnalyticsEventsDirective
                    ],
                    exports: [
                        GoogleAnayticsComponent,
                        GoogleAnalyticsEventsDirective
                    ]
                },] }
    ];
    /** @nocollapse */
    NgxGtagModule.ctorParameters = function () { return [
        { type: NgxGtagModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    return NgxGtagModule;
}());
export { NgxGtagModule };
