// ngx-gtag/esm5/lib/google-anaytics/google-anaytics.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser, DOCUMENT } from '@angular/common';
import { Router, NavigationEnd } from '@angular/router';
import { filter } from 'rxjs/operators';
import { NGX_GTAG_CONFIG } from '../google-analytics.token';
import * as i0 from "@angular/core";
import * as i1 from "@angular/router";
import * as i2 from "../google-analytics.token";
import * as i3 from "@angular/common";
var GoogleAnayticsService = /** @class */ (function () {
    function GoogleAnayticsService(router, config, document, platform) {
        this.router = router;
        this.config = config;
        this.document = document;
        this.platform = platform;
    }
    /**
     * @return {?}
     */
    GoogleAnayticsService.prototype.startTracking = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (isPlatformBrowser(this.platform)) {
            if (Array.isArray(this.config)) {
                this.config.forEach(function (conf) {
                    conf.options = {};
                    /** @type {?} */
                    var s1 = _this.renderer.createElement('script');
                    s1.src = "https://www.googletagmanager.com/gtag/js?id='" + conf.trackingId + "'";
                    _this.renderer.appendChild(_this.document.head, s1);
                });
            }
            else if (this.config.options === undefined) {
                this.config.options = {};
                /** @type {?} */
                var s1 = this.renderer.createElement('script');
                s1.src = "https://www.googletagmanager.com/gtag/js?id='" + this.config.trackingId + "'";
                this.renderer.appendChild(this.document.head, s1);
            }
            /** @type {?} */
            var s2 = this.renderer.createElement('script');
            /** @type {?} */
            var text = this.renderer.createText("window.dataLayer=window.dataLayer||[];function gtag(){dataLayer.push(arguments);}gtag('js',new Date());");
            this.renderer.appendChild(s2, text);
            this.renderer.appendChild(this.document.head, s2);
            this.router.events
                .pipe(filter(function (event) { return event instanceof NavigationEnd; }))
                .subscribe(function (navEnd) {
                if (Array.isArray(_this.config)) {
                    _this.config.forEach(function (conf) {
                        conf.options.page_path = navEnd.urlAfterRedirects;
                        gtag('config', conf.trackingId, conf.options);
                    });
                }
                else {
                    _this.config.options.page_path = navEnd.urlAfterRedirects;
                    gtag('config', _this.config.trackingId, _this.config.options);
                }
            });
        }
    };
    GoogleAnayticsService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    GoogleAnayticsService.ctorParameters = function () { return [
        { type: Router },
        { type: undefined, decorators: [{ type: Inject, args: [NGX_GTAG_CONFIG,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
    ]; };
    /** @nocollapse */ GoogleAnayticsService.ngInjectableDef = i0.defineInjectable({ factory: function GoogleAnayticsService_Factory() { return new GoogleAnayticsService(i0.inject(i1.Router), i0.inject(i2.NGX_GTAG_CONFIG), i0.inject(i3.DOCUMENT), i0.inject(i0.PLATFORM_ID)); }, token: GoogleAnayticsService, providedIn: "root" });
    return GoogleAnayticsService;
}());
export { GoogleAnayticsService };
if (false) {
    /** @type {?} */
    GoogleAnayticsService.prototype.trackingId;
    /** @type {?} */
    GoogleAnayticsService.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.router;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.config;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.document;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsService.prototype.platform;
}
