// ngx-gtag/esm5/lib/google-anaytics/google-anaytics.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Renderer2 } from '@angular/core';
import { GoogleAnayticsService } from './google-anaytics.service';
var GoogleAnayticsComponent = /** @class */ (function () {
    function GoogleAnayticsComponent(service, renderer) {
        this.service = service;
        this.renderer = renderer;
        this.service.renderer = this.renderer;
    }
    /**
     * @return {?}
     */
    GoogleAnayticsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.service.startTracking();
    };
    GoogleAnayticsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngx-gtag',
                    template: ''
                }] }
    ];
    /** @nocollapse */
    GoogleAnayticsComponent.ctorParameters = function () { return [
        { type: GoogleAnayticsService },
        { type: Renderer2 }
    ]; };
    return GoogleAnayticsComponent;
}());
export { GoogleAnayticsComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsComponent.prototype.service;
    /**
     * @type {?}
     * @private
     */
    GoogleAnayticsComponent.prototype.renderer;
}
